File: src/test/java/org/hisp/dhis/expression/UnaryOperatorExpressionTest.java
package org.hisp.dhis.expression;

import org.hisp.dhis.expression.ast.UnaryOperator;
import org.hisp.dhis.expression.spi.IllegalExpressionException;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

/**
 * Tests the {@link UnaryOperator} evaluation.
 */
class UnaryOperatorExpressionTest {

    @Test
    void testUnaryNot() {
        assertEquals(true, evaluate("!false"));
        assertEquals(false, evaluate("!true"));
        assertEquals(false, evaluate("not (true or false)"));
        assertEquals(true, evaluate("not(true and false)"));
    }

    @Test
    void testUnaryPlus() {
        assertEquals(10.0d, evaluate("++10"));
        assertEquals(8.0d, evaluate("+(+10-2)"));
    }

    @Test
    void testUnaryMinus() {
        assertEquals(-10.0d, evaluate("-10"));
        assertEquals(10.0d, evaluate("--10"));
        assertEquals(-8.0d, evaluate("-(+10-2)"));
    }

    @Test
    void testUnaryDistinct() {
        IllegalExpressionException ex = assertThrows(IllegalExpressionException.class, () -> evaluate("distinct 1"));
        assertEquals("Unary operator not supported for direct evaluation: DISTINCT", ex.getMessage());
    }

    private static Object evaluate(String expression) {
        return new Expression(expression).evaluate();
    }
}

File: src/jvmTest/java/org/hisp/dhis/lib/expression/UnaryOperatorExpressionTest.kt
package org.hisp.dhis.lib.expression

import org.hisp.dhis.lib.expression.spi.IllegalExpressionException
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith

/**
 * Tests the [UnaryOperator] evaluation.
 */
internal class UnaryOperatorExpressionTest {
    @Test
    fun testUnaryNot() {
        assertEquals(true, evaluate("!false"))
        assertEquals(false, evaluate("!true"))
        assertEquals(false, evaluate("not (true or false)"))
        assertEquals(true, evaluate("not(true and false)"))
    }

    @Test
    fun testUnaryPlus() {
        assertEquals(10.0, evaluate("++10"))
        assertEquals(8.0, evaluate("+(+10-2)"))
    }

    @Test
    fun testUnaryMinus() {
        assertEquals(-10.0, evaluate("-10"))
        assertEquals(10.0, evaluate("--10"))
        assertEquals(-8.0, evaluate("-(+10-2)"))
    }

    @Test
    fun testUnaryDistinct() {
        val ex = assertFailsWith(IllegalExpressionException::class) { evaluate("distinct 1") }
        assertEquals("Unary operator not supported for direct evaluation: DISTINCT", ex.message)
    }

    companion object {
        private fun evaluate(expression: String): Any? {
            return Expression(expression).evaluate()
        }
    }
}
